programs/vote/src/vote_state/mod.rs

```diff
@@ -226,6 +280,182 @@ impl VoteStateUpdate {
+    /// Part of the proposed tower, votes with confirmation_count > 15
+    #[serde(with = "short_vec")]
+    pub lockouts_32: Vec<CompactLockout<u32>>,
+    /// Part of the proposed tower, votes with 15 >= confirmation_count > 7
+    #[serde(with = "short_vec")]
+    pub lockouts_16: Vec<CompactLockout<u16>>,
+    /// Part of the proposed tower, votes with 7 >= confirmation_count
+    #[serde(with = "short_vec")]
+    pub lockouts_8: Vec<CompactLockout<u8>>,
+
+    /// Signature of the bank's state at the last slot
+    pub hash: Hash,
+    /// Processing timestamp of last slot
+    pub timestamp: Option<UnixTimestamp>,
+}
+
+impl From<Vec<(Slot, u32)>> for CompactVoteStateUpdate {
+    fn from(recent_slots: Vec<(Slot, u32)>) -> Self {
+        let lockouts: VecDeque<Lockout> = recent_slots
+            .into_iter()
+            .map(|(slot, confirmation_count)| Lockout {
+                slot,
+                confirmation_count,
+            })
+            .collect();
+        Self::new(lockouts, None, Hash::default())
+    }
+}
+
+impl CompactVoteStateUpdate {
+    pub fn new(mut lockouts: VecDeque<Lockout>, root: Option<Slot>, hash: Hash) -> Self {
+        let mut cur_slot = root.unwrap_or(0u64);
+        let mut cur_confirmation_count = 0;
+        let offset = lockouts
+            .pop_front()
+            .map(
+                |Lockout {
+                     slot,
+                     confirmation_count,
+                 }| {
+                    assert!(confirmation_count < 32);
+
+                    let offset = slot - cur_slot;
+                    cur_slot = slot;
+                    cur_confirmation_count = confirmation_count;
+                    offset
+                },
+            )
+            .expect("Tower should not be empty");
+        let mut lockouts_32 = Vec::new();
+        let mut lockouts_16 = Vec::new();
+        let mut lockouts_8 = Vec::new();
+
+        for Lockout {
+            slot,
+            confirmation_count,
+        } in lockouts
+        {
+            assert!(confirmation_count < 32);
+            let offset = slot - cur_slot;
+            if cur_confirmation_count >= 15 {
+                lockouts_32.push(CompactLockout {
+                    offset: offset.try_into().unwrap(),
+                    confirmation_count: cur_confirmation_count.try_into().unwrap(),
+                });
+            } else if cur_confirmation_count >= 7 {
+                lockouts_16.push(CompactLockout {
+                    offset: offset.try_into().unwrap(),
+                    confirmation_count: cur_confirmation_count.try_into().unwrap(),
+                });
+            } else {
+                lockouts_8.push(CompactLockout {
+                    offset: offset.try_into().unwrap(),
+                    confirmation_count: cur_confirmation_count.try_into().unwrap(),
+                })
+            }
+
+            cur_slot = slot;
+            cur_confirmation_count = confirmation_count;
+        }
```

Comment:
This loop iterates the slots from the smallest slot to the largest slot, so `let offset = slot - cur_slot;` is the distance from the current slot to the previous slot right?

Doesn't this have the same issue as before?

I could have `0 1 2 3 4 5 6 7 8 9 300`, and the tower would look like `0, 1, 300`

Reply:
Seems much more reasonable to just use the smallest possible representation for offsets whenever possible, rather than basing it on confirmation count

Reply:
this is computing the offset to the next slot but for the current confirmation level and slot. So in this case when we do 300 - 1 = 299, this is stored for slot 1 which has a conf count high enough to hold it.

Reply:
Hmm I see, so the datatype of the `offset` just has to be large enough to hold `2^confirmation_count` for the current confirmation count. Isn't there an off by one error here though? the largest these types can hold are `2^n - 1`

Reply:
Yes exactly. 
> Isn't there an off by one error here though? the largest these types can hold are 2^n - 1
That's why I use 15 and 7 as the partition cutoffs rather than 16 and 8. Although these `<=` could be `<` since `2^15` fits in a `u16`

Reply:
cool, sounds good to me., thanks for the explainer. Could we add some more extensive test cases testing:
1.  these border conditions
2.  Various offsets 2^8, 2^16, 2^32 for instance to test each of those partitions